### DISCUSSION TITLE
How to reset OpenIdConnectConfiguration when IssuerSigningKeys changes?

### DISCUSSION THREAD

**Discussion by oledid** [dotnet/aspnetcore#27067] [Q&A] [answered]
Hi

I have previously asked this question on [stackoverflow](https://stackoverflow.com/questions/64287179/how-to-reset-openidconnectconfiguration-when-issuersigningkeys-changes), but I haven't gotten an answer.

```csharp
.AddOpenIdConnect(OpenIdConnectDefaults.AuthenticationScheme, options =>
            {
                var configurationManager = new ConfigurationManager<OpenIdConnectConfiguration>("https://accounts.google.com/.well-known/openid-configuration", new OpenIdConnectConfigurationRetriever());
                options.Configuration = configurationManager.GetConfigurationAsync().Result;
```

```csharp
options.TokenValidationParameters = new TokenValidationParameters
                {
                    NameClaimType = "name",
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKeys = options.Configuration.SigningKeys
                };
```

Provider changes their keys:
```
System.Exception: An error was encountered while handling the remote login.
 ---> Microsoft.IdentityModel.Tokens.SecurityTokenSignatureKeyNotFoundException: IDX10501: Signature validation failed. Unable to match key: 
kid: '[PII is hidden. For more details, see https://aka.ms/IdentityModel/PII.]'.
```

With an app that is always running, is there a way to refresh the info from the .well-known when the OpenID Connect provider changes their certs? Right now I'm stuck with resetting the azure app service when it happens.

**Comment by Tratcher** [comment, marked as answer]
SecurityTokenSignatureKeyNotFoundException will automatically trigger a refresh from the ConfigurationManager.
https://github.com/dotnet/aspnetcore/blob/f54b590804ecacc328d9050ee1bdbaede9fb57a1/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs#L765-L772

Why are you forcibly resolving the keys up front from the ConfigurationManager rather than passing the configuration manager to OpenIdConnectOptions?
https://github.com/dotnet/aspnetcore/blob/f54b590804ecacc328d9050ee1bdbaede9fb57a1/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs#L134

**Comment by oledid** [comment]
Thanks. I must have misunderstood how to use the api.

Are you saying that this should sove my problem?

```csharp
.AddOpenIdConnect(OpenIdConnectDefaults.AuthenticationScheme, options =>
			{
				var configurationManager = new ConfigurationManager<OpenIdConnectConfiguration>(Configuration["Google:WellKnown"], new OpenIdConnectConfigurationRetriever());
				//options.Configuration = configurationManager.GetConfigurationAsync().Result; // removed
				options.ConfigurationManager = configurationManager; // added
```

Edit: Or I can just set `options.MetadataAddress`, which will create a `ConfigurationManager`. Thanks!

**Reply by Tratcher** [reply]
Yeah, there are a lot of properties there to allow for customization, but MetadataAddress is all you need for the basic scenario.
